### ISSUE TITLE
Model Zoo

### ISSUE THREAD

**Issue by simon-jonas-buehrer** [aplesner/difflut#24]
## Problem Statement

DiffLUT lacks a centralized repository of ready-to-use model architectures and pre-trained models, making it harder for users to quickly evaluate or benchmark the library.

## Proposed Solution

Add a `models/` folder containing:

1. **Model architectures** - Ready-to-use network definitions for common tasks.
2. **Pre-trained weights** - Checkpoints for standard benchmarks.
3. **Model factory** - Use and expand the existing `Registry` to manage models.

## Directory Structure Proposal

```
difflut/
├── encoder/
├── layers/
├── nodes/
├── utils/
└── models/              # NEW
    ├── __init__.py
    ├── base_model.py    # Base class for models
    ├── mnist.py         # MNIST classifiers
    ├── cifar.py         # CIFAR-10/100 classifiers
    ├── speech.py        # Speech command recognition
    └── pretrained/      # Pre-trained checkpoints
        ├── mnist_linear_4input.pth
        ├── cifar10_dwn_6input.pth
        └── ...
```

## Example API

```python
# in difflut/models/mnist.py
from difflut import registry

@registry.register_model('mnist_linear')
class MNISTLinearModel(nn.Module):
    def __init__(self, hidden_size=256, node_type='linear_lut'):
        # build using REGISTRY.get_encoder/get_layer/get_node as needed
        ...

# elsewhere
from difflut.registry import  get_model,list_models

print(list_models())
# Output: ['mnist_linear',...]
model = get_model('mnist_linear', pretrained=False)
```

### 1. Scope of Model Zoo

**Option A: Minimal** - Only architectures, no pre-trained weights
- Pros: Smaller maintenance burden, no storage requirements
- Cons: Less useful for quick evaluation

**Option B: Full** - Architectures + pre-trained weights for standard benchmarks
- Pros: Complete out-of-the-box experience, easier benchmarking
- Cons: Storage requirements, need hosting infrastructure

**Option C: Hybrid** - Architectures in repo, weights hosted externally (e.g., HuggingFace)
- Pros: Clean separation, leverages existing infrastructure
- Cons: External dependency

### 2. Model Organization

**By Dataset:**
```
models/mnist.py
models/cifar.py
models/speech_commands.py
```

**By Architecture Type:**
```
models/classification.py
models/regression.py
models/autoencoder.py
```

**Flat Structure:**
```
models/mnist_linear.py
models/mnist_dwn.py
models/cifar10_linear.py
```

### 3. Configuration Management

**Hardcoded in Python:**
```python
class MNISTLinear(nn.Module):
    def __init__(self):
        super().__init__()
        self.encoder = ThermometerEncoder(num_bits=8)
        self.layer1 = RandomLayer(input_size=784*8, output_size=128, ...)
```

**YAML/JSON Configs:**
```yaml
# models/configs/mnist_linear.yaml
encoder:
  type: thermometer
  num_bits: 8
layers:
  - type: random
    output_size: 128
    node_type: linear
```

**Registry-based with NodeConfig/LayerConfig:**
```python
config = {
    'encoder': EncoderConfig(type='thermometer', num_bits=8),
    'layers': [LayerConfig(type='random', output_size=128, ...)]
}
```

### 4. Pre-trained Weight Management

- Should we include weights in git repository (increases repo size)?
- Host on external service (HuggingFace Hub, Google Drive, institutional storage)?
- Implement automatic download on first use?

### ModelConfig
Similar to NodeConfig, LayerConfig

### Testing Requirements

- Model instantiation tests
- Forward pass dimension validation
- Pre-trained weight loading tests
- Inference accuracy tests (if pre-trained)
- Cross-version compatibility tests

## References

- Related to: Issue #22 (FPGA Export) - Models should be exportable
- Depends on: Current encoder/layer/node infrastructure (stable)

---

**Requires Discussion**
